pkg/registry/common/memory/nse_server_test.go

```diff
@@ -177,6 +183,180 @@ func TestNetworkServiceEndpointRegistryServer_RegisterAndFindByLabelWatch(t *tes
 	require.True(t, proto.Equal(expected, <-ch))
 }
 
+func TestNetworkServiceEndpointRegistryServer_DataRace(t *testing.T) {
+	defer goleak.VerifyNone(t, goleak.IgnoreCurrent())
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	s := memory.NewNetworkServiceEndpointRegistryServer()
+
+	go func() {
+		for ctx.Err() == nil {
+			nse, err := s.Register(ctx, &registry.NetworkServiceEndpoint{Name: "nse-1"})
+			require.NoError(t, err)
+
+			_, err = s.Unregister(ctx, nse)
+			require.NoError(t, err)
+		}
+	}()
+
+	c := adapters.NetworkServiceEndpointServerToClient(s)
+
+	wg := new(sync.WaitGroup)
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			stream, err := c.Find(ctx, &registry.NetworkServiceEndpointQuery{
+				NetworkServiceEndpoint: &registry.NetworkServiceEndpoint{Name: "nse-1"},
+				Watch:                  true,
+			})
+			assert.NoError(t, err)
+
+			for j := 0; j < 100; j++ {
+				nse, err := stream.Recv()
+				assert.NoError(t, err)
+
+				nse.Name = ""
+			}
+		}()
+	}
+
+	wg.Wait()
+}
+
+func TestNetworkServiceEndpointRegistryServer_SlowReceiver(t *testing.T) {
```

Comment:
This test is invalid.

You are trying to write 100 entries to a channel that has a limited 10 by default. Please look https://github.com/networkservicemesh/sdk/blob/master/pkg/registry/common/memory/common.go#L19
And check how the adapted client is actually working.

Please use option setChannelSize(100) or
Rework this test to correctly read/write from the channel to simulate the real client/server.

Reply:
It is actually very similar to the `discover` server using watching `Find` and finishing receiving after the first match.